File: code/src/application/model/game/GameLibrary.java
package application.model.game;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Manages a collection of games.
 * 
 * @author Thomas Lamont
 * @version Sprint 1
 */
public class GameLibrary {

	private List<Game> games;

	/**
	 * Constructs a new GameLibrary instance.
	 */
	public GameLibrary() {
		this.games = new ArrayList<>();
	}

	/**
	 * Adds a game to the library.
	 * 
	 * @param game The game to add.
	 */
	public void addGame(Game game) {
		if (game == null) {
			throw new IllegalArgumentException("Game must not be null");
		}
		for (Game existingGame : this.games) {
			if (existingGame.equals(game)) {
				throw new IllegalArgumentException("Duplicate game ID: " + game.getGameID());
			}
		}
		this.games.add(game);
	}

	/**
	 * Removes a game from the library by its game ID.
	 * 
	 * @param gameID The ID of the game to remove.
	 * @return true if the game was found and removed, false otherwise.
	 */
	public boolean removeGame(int gameID) {
		return this.games.removeIf(game -> game.getGameID() == gameID);
	}

	/**
	 * Gets the list of all games in the library.
	 * 
	 * @return A list of all games.
	 */
	public List<Game> getGames() {
		return new ArrayList<>(this.games);
	}

	/**
	 * Finds a game by its ID.
	 * 
	 * @param gameID The ID of the game to find.
	 * @return An Optional containing the found game or empty if not found.
	 */
	public Optional<Game> findGameById(int gameID) {
		return this.games.stream().filter(game -> game.getGameID() == gameID).findFirst();
	}

	@Override
	public String toString() {
		return this.games.stream().map(Game::toString).collect(Collectors.joining("\n"));
	}
}

File: code/server-side-impl/src/src/model/games/gamelibrary.py
'''
Created on Mar 2, 2024

@author: Thomas
'''
class GameLibrary:
    """
    A class to represent a library of video games.
    
    This class provides methods to add, remove, and search for games within the library,
    allowing for the management of a collection of `Game` objects.
    
    Attributes:
        games (list[Game]): A list of games in the library.
    """
    def __init__(self):
        """
        Initializes a new instance of the GameLibrary class, initially empty.
        """
        self.games = []

    def add_game(self, game):
        """
        Adds a new game to the library.
        
        Args:
            game (Game): The game to add to the library.
            
        Raises:
            ValueError: If the game is None or if a game with the same ID already exists in the library.
        """
        if game is None:
            raise ValueError("Game must not be null")
        if any(existing_game.game_id == game.game_id for existing_game in self.games):
            raise ValueError(f"Duplicate game ID: {game.game_id}")
        self.games.append(game)
        
    def get_games(self):
        """
        Returns a list of all games in the library.
        
        Returns:
            list[Game]: The list of games in the library.
        """
        return self.games

    def remove_game(self, game_id):
        """
        Removes a game from the library by its unique game ID.
        
        Args:
            game_id (int): The ID of the game to remove.
        """
        self.games = [game for game in self.games if game.game_id != game_id]
    
    def remove_all_games(self):
        """
        Clears the library of all games.
        """
        self.games = []

    def find_game_by_id(self, game_id):
        """
        Searches for a game by its ID and returns it if found.
        
        Args:
            game_id (int): The ID of the game to search for.
            
        Returns:
            Game: The game with the matching ID, or None if no match is found.
        """
        for game in self.games:
            if game.game_id == game_id:
                return game
        return None